snuba/utils/streams/synchronized.py

```diff
@@ -0,0 +1,203 @@
+from dataclasses import dataclass
+from threading import Event
+from typing import Callable, Mapping, MutableMapping, Optional, Sequence, Set
+
+from snuba.utils.concurrent import Synchronized, execute
+from snuba.utils.streams.consumer import Consumer, ConsumerError, EndOfPartition
+from snuba.utils.streams.types import Message, Partition, Topic, TPayload
+
+
+@dataclass(frozen=True)
+class Commit:
+    __slots__ = ["group", "partition", "offset"]
+
+    group: str
+    partition: Partition
+    offset: int
+
+
+class SynchronizedConsumer(Consumer[TPayload]):
+    def __init__(
+        self,
+        consumer: Consumer[TPayload],
+        commit_log_consumer: Consumer[Commit],
+        commit_log_topic: Topic,
+        commit_log_groups: Set[str],
+    ) -> None:
+        self.__consumer = consumer
+
+        self.__commit_log_consumer = commit_log_consumer
+        self.__commit_log_topic = commit_log_topic
+        self.__commit_log_groups = commit_log_groups
+
+        self.__remote_offsets: Synchronized[
+            Mapping[str, MutableMapping[Partition, int]]
+        ] = Synchronized({group: {} for group in commit_log_groups})
+
+        self.__commit_log_worker_stop_requested = Event()
+        self.__commit_log_worker = execute(self.__run_commit_log_worker)
+
+        # The set of partitions that have been paused by the caller/user. This
+        # takes precedence over whether or not the partition should be paused
+        # due to offset synchronization.
+        self.__paused: Set[Partition] = set()
+
+    def __run_commit_log_worker(self) -> None:
+        # TODO: This needs to roll back to the initial offset.
+
+        # TODO: This needs to ensure that it is subscribed to all partitions.
+
+        self.__commit_log_consumer.subscribe([self.__commit_log_topic])
+
+        while not self.__commit_log_worker_stop_requested.is_set():
+            try:
+                message = self.__commit_log_consumer.poll(0.1)
```

Comment:
Do we ever have to commit the offset we consumed from the commit log ? Or do we plan to always restart from the beginning of the topic?

Reply:
This needs to restart from the beginning of the topic to ensure it has the latest available offset for all partitions in the case scenario where one or more partitions is not being actively committed to but the local offset is before the remote offset, which would allow the local consumer to do work.

I don't think that is a very common scenario (so the impact of resetting to the latest offset at startup is honestly pretty low) but I figured it's better to play it safe here. With topic compaction there shouldn't be that much of a backlog to burn through — but I don't have any actual metrics about how long it actually takes the consumer in Sentry to do that today. It doesn't seem to lag very far behind on deploys, so it can't be _that_ long, I guess?